File: shortest/maze.cpp
#include <queue>
#include <complex>
#include <iostream>
#include <fstream>

#define MAX_W 30
#define MAX_H 30
using namespace std;
using point = complex<int>;
point directions[] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

int canGoX[MAX_H + 2][MAX_W + 2];
int canGoY[MAX_H + 2][MAX_W + 2];
int steps[MAX_H][MAX_W] = {};

int canGo(point & from, point & direct) {
  if (direct.imag() == 0) {
    if (direct.real() > 0)
      return canGoX[1+ from.imag()][1 + from.real()];
    else
      return canGoX[1 + from.imag()][1 + from.real() - 1];
  } else { /* direct.real()==0*/
    if (direct.imag() > 0)
      return canGoY[1 + from.imag()][1 + from.real()];
    else
      return canGoY[1 + from.imag() - 1][1 + from.real()];
  }
}

void setupBoard(int w, int h) {
  int i = 0;
  for (i = 0; i < h + 2; i++) {
    int j;
    for (j = 0; j < w + 2; j++) {
      canGoX[i][j] = 0;
      canGoY[i][j] = 0;
    }
  }
  i = 0;
  while (1) {
    int j;
    for (j = 0; j < w - 1; j++) {
      int d;
      cin >> d;
      canGoX[1 + i][1 + j] = 1 - d;
    }
    if (i + 1 == h) break;
    for (j = 0; j < w; j++) {
      int d;
      cin >> d;
      canGoY[1 + i][1 + j] = 1 - d;
    }
    i++;
  }
}
int clearSteps(int w, int h) { 
  for (int i = 0; i < h; i++) {
    for (int j = 0; j < w; j++) {
      steps[i][j] = 0;
    }
  }
}
int getStep(point p) { return steps[p.imag()][p.real()]; }
void setStep(point p, int v) { steps[p.imag()][p.real()] = v; }

int solve(int w, int h) {
  point start = {0, 0};
  point goal = {w - 1, h - 1};
  queue<point> q;
  q.push(start);
  clearSteps(w, h);
  setStep(start, 1);

  while (!q.empty()) {
    point here = q.front();
    q.pop();
    int currentStep = steps[here.imag()][here.real()];
    for (point & d: directions) {
      if (canGo(here, d)) {
        point next = here + d;
        if (getStep(next) > 0) continue;
        if (next == goal) return currentStep + 1;
        setStep(next, currentStep + 1);
        q.push(next);
      }
    }
  }
  return 0;
}


int main() {
  ifstream ifs("mazeData/B1.in");
  if(ifs.good()) {// file があるときはそちらから
    cin.rdbuf(ifs.rdbuf());
  }
  while (1) {
    int w, h, result, ans;
    cin >> w >> h;
    if (w == 0 && h == 0) break;
    setupBoard(w, h);
    result = solve(w, h);
    cout << result << endl;
  }
  return 0;
}

File: shortest/maze.py
from collections import deque
#ファイルから読み込みたい場合は、直後の行を有効化してください。完成時は、コメントに戻すのを忘れずに！
#import sys; sys.stdin =  open('mazeData/B1.in', 'r')
directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]


def setup():
    i = 0
    while True:
        wall = input().split()
        for j, d in enumerate(wall):
            if d == "0":
                canGoX[1 + i][1 + j] = True
        if i + 1 == h:
            break
        wall = input().split()
        for j, d in enumerate(wall):
            if d == "0":
                canGoY[1 + i][1 + j] = True
        i += 1


def can_go(x, y, dx, dy) -> bool:
    if dy == 0:
        if dx > 0:
            return canGoX[1 + y][1 + x]
        else:
            return canGoX[1 + y][1 + x - 1]
    else:
        if dy > 0:
            return canGoY[1 + y][1 + x]
        else:
            return canGoY[1 + y - 1][1 + x]


def solve() -> int:
    goal = [w - 1, h - 1]
    q = deque([[0, 0]])
    steps = [[0 for _ in range(w)] for _ in range(h)]
    steps[0][0] = 1

    while q:
        x, y = q.popleft()
        current_step = steps[y][x]
        for dx, dy in directions:
            if can_go(x, y, dx, dy):
                nx, ny = [x + dx, y + dy]
                if steps[ny][nx] > 0:
                    continue
                if [nx, ny] == goal:
                    return current_step + 1
                steps[ny][nx] = current_step + 1
                q.append([nx, ny])
    return 0


while True:
    w, h = map(int, input().split())
    if w == h == 0:
        exit(0)
    canGoX = [[False for _ in range(w + 2)] for _ in range(h + 2)]
    canGoY = [[False for _ in range(w + 2)] for _ in range(h + 2)]
    setup()
    result = solve()
    print(result)
